### PR TITLE
Solarpanels

### PR DESCRIPTION
**Pull request by GermanHydrogen**
Added power generation to rugged and flexible solar panels.

Closing #145 
Closing #3

### CODE REVIEW

**Comment by GermanHydrogen** [comment]
@y0014984 Is it possible to move the ace interaction points for the animations?
The interaction point of the yaw axis of the rugged solar panel is very close to the objects main interaction point.

**Comment by y0014984** [comment]
> @y0014984 Is it possible to move the ace interaction points for the animations? The interaction point of the yaw axis of the rugged solar panel is very close to the objects main interaction point.

We are using Selection Names for dynamic interaction positions. That makes sense for the rugged lights, because these selection positions use dynamically the actual hight of a lamp. But for other equipment this doesn't fit very well. Therefore we have two options:

Option A: Use the Main Interaction Point
Option B: Define a custom iteraction position that is different from the main interaction point

@GermanHydrogen please choose your prefered option and assign me a ticket. I will make the adjustment as soon as I can.

**Comment by y0014984** [comment]
I will do the review on friday.

**Comment by y0014984** [comment]
@GermanHydrogen : I tested the solar panel output, especially the modification by visibility of the sun. In full shadow behind a building the panel produced full power output. So I checked the code myself and it seems that the virtual position of the sun is horizontally approx. 90 degrees off it's current position. I tried that on altis and tanoa in March and November. It's mostly 90 degrees offset to the east, see screenshots. In October on Altis the Offset was more something like 150 degrees.

https://i.imgur.com/0P9AH1P.jpg
https://i.imgur.com/Nl6PFHp.jpg

Additionally you can see that the calculated distance for the virtual sun is only about 5-10 meters away from the solar panel. A higher Value would be better in the vectorMultiply function so the virtual sun is something like 50 to 100 meters away. You use the unmodified position of the solar panel to calculate the position of the virtual sun, but you shoud use the modified (lifted) position for exact results, see screenshot.

You could interchange lines 29 and 30 and modifiy the new line 30 to look like

```sqf
private _solPos = _panelPos vectorAdd (_solVec vectorMultiply 10);
```

File: addons/power/CfgVehicles.hpp [hunk 1]
```diff
@@ -196,13 +196,82 @@ class CfgVehicles
...
+			defaultPowerLevel = 0;
+
+			turnOnAction = "_this call AE3_power_fnc_turnOnSolarAction";
+			turnOffAction = "_this call AE3_power_fnc_turnOffSolarAction";
+
+			class AE3_SolarGenerator
+			{
+				powerMax = 0.1/3600; // In this case per panel
+				orientationFnc = "_this call AE3_power_fnc_multSolarPanelOrientation";
+				height = 1.5;
```
Comment (y0014984) [inline addons/power/CfgVehicles.hpp:219]: 1.2 fits better

File: addons/power/functions/fnc_solarCalculation.sqf [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+{
+	_buffer = _x vectorCos _solVec;
+	if (_buffer > 0) then 
+	{
+		_mult = _mult + _buffer;
+	};
+}forEach _normalList;
+
+// If sun is visible
+private _solPos = getPosASL _solar vectorAdd (_solVec vectorMultiply 10);
```
Comment (y0014984) [inline addons/power/functions/fnc_solarCalculation.sqf:29]: //Move down one line and modify likes this:
private _solPos = _panelPos vectorAdd (_solVec vectorMultiply 10);

Comment (y0014984) [reply addons/power/functions/fnc_solarCalculation.sqf:29]: vectorMultiplay 100 would be better

File: addons/power/functions/fnc_turnOffSolarAction.sqf [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+		{
+
+		},
+		{},
+		("Turn Off")
+	] call ace_common_fnc_progressBar;
+};
+
+[_entity, true, [0, 1, 0], 0] call ace_dragging_fnc_setDraggable;
+
```
Comment (y0014984) [inline addons/power/functions/fnc_turnOffSolarAction.sqf:27]: //We should add something like this, so the checkPowerGenAction will return 0 Watts if the solar panel is powered down.

_entity setVariable ['AE3_power_powerCapacity', 0, true];

**Review by y0014984 — COMMENTED** [review]
Besides my comments I have no additional things to say. If these things get fixed then you can merge this PR.

**Review by y0014984 — COMMENTED** [review]
I have tested the sun calculations on 17 maps, official and custom ones. (Altis, Stratis, Malden, Livonia, Tanoa, Virtual Reality, Cam Lao Nam, Khe San, Weferlingen, Sefron Ramal, Chernarus, Kujari, Lythium, Rosche, Ruha, Takistan, Virolahti). On all maps the sun calculations worked very well. Sometimes the virtual sun seems to be a little bit too high but thats not a big deal. Only on tanoa the offsets were bigger. Still usable but i qould say a few degrees offset. Perhaps this is due to the position of tanoa, that is very close to the equator. To sum it all together I would say we have very good sun calculations to work with.

The power production on the other hand is a problem as long as we still use the sunOrMoon command to calculate the power output instead of the sun angle and visablitity of the sun. The solar panel still produces power (loads it's internal battery) in very dark (not night) conditions or in full shadow behind buildings. If you check Power Output, it says 0 Watts, but the battery is still loading. @GermanHydrogen is this the intended behavior?

**Comment by y0014984** [comment]
Looks good now. I will merge.

**Pull request merged**